iroh-net/src/discovery/static_discovery.rs

```diff
@@ -0,0 +1,126 @@
+    /// # #[derive(Default)] struct Args { tickets: Vec<NodeTicket> }
+    /// # let args = Args::default();
+    /// // get tickets from command line args
+    /// let tickets: Vec<NodeTicket> = args.tickets;
+    /// // create a StaticDiscovery from the tickets. Ticket info will be combined if multiple tickets refer to the same node.
+    /// let discovery = StaticDiscovery::from_node_addrs(tickets);
+    /// // create an endpoint with the discovery
+    /// let endpoint = Endpoint::builder()
+    ///     .add_discovery(|_| Some(discovery))
+    ///     .bind().await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn from_node_addrs(infos: impl IntoIterator<Item = impl Into<NodeAddr>>) -> Self {
+        let res = Self::default();
+        for info in infos {
+            res.add_node_addr(info);
+        }
+        res
+    }
+
+    /// Add node info for the given node id.
+    ///
+    /// This will completely overwrite any existing info for the node.
+    pub fn set_node_addr(&self, info: impl Into<NodeAddr>) {
+        let info: NodeAddr = info.into();
+        let mut guard = self.nodes.write().unwrap();
+        guard.insert(info.node_id, info.info);
+    }
+
+    /// Add node info for the given node id, combining it with any existing info.
+    ///
+    /// This will add any new direct addresses and overwrite the relay url.
+    pub fn add_node_addr(&self, info: impl Into<NodeAddr>) {
+        let info: NodeAddr = info.into();
+        let mut guard = self.nodes.write().unwrap();
+        match guard.entry(info.node_id) {
+            Entry::Occupied(mut entry) => {
+                let existing = entry.get_mut();
+                existing.direct_addresses.extend(info.info.direct_addresses);
+                existing.relay_url = info.info.relay_url;
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(info.info);
+            }
+        }
+    }
+
+    /// Get node info for the given node id.
+    pub fn get_node_addr(&self, node_id: NodeId) -> Option<NodeAddr> {
+        let guard = self.nodes.read().unwrap();
+        let info = guard.get(&node_id).cloned()?;
+        Some(NodeAddr { node_id, info })
+    }
+
+    /// Remove node info for the given node id.
+    pub fn remove_node_addr(&self, node_id: NodeId) {
+        let mut guard = self.nodes.write().unwrap();
+        guard.remove(&node_id);
+    }
+}
+
+impl Discovery for StaticDiscovery {
+    fn publish(&self, _info: &AddrInfo) {}
+
+    fn resolve(
+        &self,
+        _endpoint: crate::Endpoint,
+        node_id: NodeId,
+    ) -> Option<futures_lite::stream::Boxed<anyhow::Result<super::DiscoveryItem>>> {
+        let guard = self.nodes.read().unwrap();
+        let info = guard.get(&node_id);
+        match info {
+            Some(addr_info) => {
+                let item = DiscoveryItem {
+                    node_id,
+                    provenance: Self::PROVENANCE,
+                    last_updated: None,
```

Comment:
you could store the timestamp of insertion, and use it here

Reply:
Not sure about that. I think we don't have a single discovery service that sets this field, and also it is unused in code. Maybe just remove it?

Reply:
@ramfox @flub thoughts on this field?

Reply:
perhaps it should be used?  quoting ramfox's code comment:

> It is only truly relevant when comparing `relay_urls`, since we can attempt to dial any number of socket addresses, but expect each node to have one "home relay" that we will attempt to contact them on. This means we would need some way to determine which relay url to choose between, if more than one relay url is reported

So I would tentatively leave this in I think.  I can imagine you adding one RelayUrl statically, later the node publishing to DNS and providing a different one.  I think that's a situation we should probably handle even if we don't right now.

Reply:
I added filling out this number, despite nobody using it as of now...